Update to atomics

eschnett opened the pull request:
- make `atomic_cas!` return the old value instead of a "success" flag
- support atomic floating-point types
- implement some atomic floating-point operations via `atomic_cas!`

The success flag previously returned by `atomic_cas!` can be unambiguously determined from the old value that is now returned. An alternative implementation would return a tuple instead.

Atomic floating-point types are trivial to implement for load, store, compare-and-swap, and xchg. Do that. Other operations (add, subtract, min, max) can be implemented via compare-and-swap and a loop.

Some of these features are not supported by older LLVM versions; these require bitcasts as a work-around.

Atomic pointers and atomic booleans could also be supported, but are not part of this patch.

tkelman commented:
How widely used will this kind of thing be *within base*? Just wondering if it doesn't need to touch the C runtime, maybe more of this should be moved to libraries.

eschnett commented:
@tkelman Good question. My guess is that, if base contains any serious multi-threaded numerical algorithms (e.g. reductions), these will come in handy. The alternative is to use locks, which is likely less efficient.

Supporting atomic floating-point types per se should be in Base, if atomic integers are supported. It would be awkward to use different types for those. Whether atomic floating-point operations should be in Base is a different question.

tkelman commented:
Is there a minimum glibc version needed for these to work? See Travis failures...

yuyichao commented:
Ref https://github.com/JuliaLang/julia/issues/14818

eschnett commented:
With 128-bit atomics disabled, this now works everywhere. (Before, 128-bit atomics weren't tested, and presumably didn't work either.)

#15219 offers one solution that would allow us to re-enable 128-bit atomics in the future.

tkelman commented:
would all intermediate commits pass here?

eschnett commented:
No they don't; I'll squash once it looks good.

eschnett commented:
Rebased and squashed. Should I push?

eschnett committed:
Update to atomics

- make atomic_cas! return the old value instead of a "success" flag
- support atomic floating-point types
- implement some atomic floating-point operations via atomic_cas!
- disable 128-bit atomic types on i686 where our LLVM install is currently broken

The success flag previously returned by atomic_cas! can be unambiguously determined from the old value that is now returned. An alternative implementation would return a tuple instead.

Atomic floating-point types are trivial to implement for load, store, compare-and-swap, and xchg. Do that. Other operations (add, subtract, min, max) can be implemented via compare-and-swap and a loop.

Some of these features are not supported by older LLVM versions; these require bitcasts as a work-around.

Atomic pointers and atomic booleans could also be supported, but are not part of this patch.

eschnett commented:
Now uses `let` in the tests.

eschnett commented:
Ping

eschnett commented:
@tkelman Does this look good now?

tkelman commented:
I guess so? It's a fair amount of additions to undocumented code that I don't entirely understand the utility of having in Base right now.

yuyichao commented:
LGTM in general. I'm wondering how useful the atomic floating point arithmetics are though. Do people often need a synchronization after every floating point operation or is it more common to sync (with a `cas`) after completing all operations.

eschnett commented:
Most of the changes are code cleanup, modernization, and corrections, as well as a necessary slight change of semantics (stricter ordering). The actual addition is only 12 lines for floating-point operations; if you don't like those, we can take them out, although I find them useful. There's also quite a number of tests added, but these address mostly old (previously untested) functionality.

There are algorithms where each thread contributes one floating-point number to a result (e.g. sum, maximum), e.g. as part of a parallel reduction operation. This is most easily handled as a single atomic operation per thread, after each thread performed the reduction internally on many numbers. Alternatives are e.g. taking a lock, implementing this manually via `atomic_cas!`, or implementing a message passing dataflow between the threads, or serializing the final step of the reduction. I think it's a small addition for a large benefit to parallel programmers.

Cray has (well, used to have...) hardware support for some atomic floating-point operations, and if this was supported by LLVM, the respective LLVM intrinsics would handle both integer and floating point numbers, so that these 12 lines could be combined with the respective lines above handling integers, without changing the user-facing API.

StefanKarpinski commented:
I think this is fine as long as we continue to make it crystal clear that for now all threading-related APIs are strictly experimental and may change drastically (including being moved into packages) before we settle on an official API.

eschnett commented:
@StefanKarpinski I very much hope so, because there's quite a few rough corners in the current API that I want to see smoothed.

tkelman commented:
And hopefully documented, eventually.

eschnett commented:
Challenge accepted.
